webclient/grpc/src/main/java/io/helidon/webclient/grpc/GrpcClientCall.java

```diff
@@ -0,0 +1,191 @@
+        headerData.writeInt8(0);                                // no compression
+        headerData.writeUnsignedInt32(messageData.available());         // length prefixed
+        sendingQueue.add(BufferData.create(headerData, messageData));
+        startWriteBarrier.countDown();
+    }
+
+    protected void startStreamingThreads() {
+        // write streaming thread
+        writeStreamFuture = executor.submit(() -> {
+            try {
+                startWriteBarrier.await();
+                LOGGER.fine("[Writing thread] started");
+
+                boolean endOfStream = false;
+                while (isRemoteOpen()) {
+                    LOGGER.finest("[Writing thread] polling sending queue");
+                    BufferData bufferData = sendingQueue.poll(WAIT_TIME_MILLIS, TimeUnit.MILLISECONDS);
+                    if (bufferData != null) {
+                        if (bufferData == EMPTY_BUFFER_DATA) {     // end marker
+                            LOGGER.finest("[Writing thread] sending queue end marker found");
+                            if (!endOfStream) {
+                                LOGGER.finest("[Writing thread] sending empty buffer to end stream");
+                                clientStream().writeData(EMPTY_BUFFER_DATA, true);
+                            }
+                            break;
+                        }
+                        endOfStream = (sendingQueue.peek() == EMPTY_BUFFER_DATA);
+                        boolean lastEndOfStream = endOfStream;
+                        LOGGER.finest(() -> "[Writing thread] writing bufferData " + lastEndOfStream);
+                        clientStream().writeData(bufferData, endOfStream);
+                    }
+                }
+            } catch (Throwable e) {
+                LOGGER.finest(e.getMessage());
+            }
+            LOGGER.fine("[Writing thread] exiting");
+        });
+
+        // read streaming thread
+        readStreamFuture = executor.submit(() -> {
+            try {
+                startReadBarrier.await();
+                LOGGER.fine("[Reading thread] started");
+
+                // read response headers
+                clientStream().readHeaders();
+
+                while (isRemoteOpen()) {
+                    // drain queue
+                    drainReceivingQueue();
+
+                    // trailers received?
+                    if (clientStream().trailers().isDone()) {
+                        LOGGER.finest("[Reading thread] trailers received");
+                        break;
+                    }
+
+                    // attempt to read and queue
+                    Http2FrameData frameData;
+                    try {
+                        frameData = clientStream().readOne(WAIT_TIME_MILLIS_DURATION);
+                    } catch (StreamTimeoutException e) {
+                        LOGGER.fine("[Reading thread] read timeout");
+                        continue;
```

Comment:
shouldn't this close this stream?

Reply:
It's currently re-trying, although we could be more draconian I suppose